src/renderer/renderScene.ts

```diff
@@ -328,468 +328,541 @@ export const _renderScene = ({
-      }),
+  // Grid
+  if (renderGrid && appState.gridSize) {
+    strokeGrid(
+      context,
+      appState.gridSize,
+      -Math.ceil(renderConfig.zoom.value / appState.gridSize) *
+        appState.gridSize +
+        (renderConfig.scrollX % appState.gridSize),
+      -Math.ceil(renderConfig.zoom.value / appState.gridSize) *
+        appState.gridSize +
+        (renderConfig.scrollY % appState.gridSize),
+      normalizedCanvasWidth / renderConfig.zoom.value,
+      normalizedCanvasHeight / renderConfig.zoom.value,
     );
+  }
 
-    let editingLinearElement: NonDeleted<ExcalidrawLinearElement> | undefined =
-      undefined;
-    visibleElements.forEach((element) => {
-      try {
-        renderElement(element, rc, context, renderConfig, appState);
-        // Getting the element using LinearElementEditor during collab mismatches version - being one head of visible elements due to
-        // ShapeCache returns empty hence making sure that we get the
-        // correct element from visible elements
-        if (appState.editingLinearElement?.elementId === element.id) {
-          if (element) {
-            editingLinearElement =
-              element as NonDeleted<ExcalidrawLinearElement>;
-          }
-        }
-        if (!isExporting) {
-          renderLinkIcon(element, context, appState);
+  // Paint visible elements
+  const visibleElements = elements.filter((element) =>
+    isVisibleElement(element, normalizedCanvasWidth, normalizedCanvasHeight, {
+      zoom: renderConfig.zoom,
+      offsetLeft: appState.offsetLeft,
+      offsetTop: appState.offsetTop,
+      scrollX: renderConfig.scrollX,
+      scrollY: renderConfig.scrollY,
+    }),
+  );
+
+  let editingLinearElement: NonDeleted<ExcalidrawLinearElement> | undefined =
+    undefined;
+  visibleElements.forEach((element) => {
+    try {
+      renderElement(element, rc, context, renderConfig, appState);
+      // Getting the element using LinearElementEditor during collab mismatches version - being one head of visible elements due to
+      // ShapeCache returns empty hence making sure that we get the
+      // correct element from visible elements
+      if (appState.editingLinearElement?.elementId === element.id) {
+        if (element) {
+          editingLinearElement = element as NonDeleted<ExcalidrawLinearElement>;
         }
-      } catch (error: any) {
-        console.error(error);
       }
-    });
+      if (!isExporting) {
+        renderLinkIcon(element, context, appState);
+      }
+    } catch (error: any) {
+      console.error(error);
+    }
+  });
 
-    if (editingLinearElement) {
-      renderLinearPointHandles(
+  if (editingLinearElement) {
+    renderLinearPointHandles(
+      context,
+      appState,
+      renderConfig,
+      editingLinearElement,
+    );
+  }
+  context.restore();
+  context.restore();
+
+  return { atLeastOneVisibleElement: visibleElements.length > 0 };
+};
+
+export const renderCanvasUI = ({
+  elements,
+  appState,
```

Comment:
renderConfig is used lot many places in both

Reply:
yep. I'd suggest creating two subtypes: `ContentRenderConfig` and `UIRenderConfig` which will both contain only the props that each render function requires. So that when we memoize, the functions don't memoize on props they don't need.

Reply:
done